cmd/controller/main.go

```diff
@@ -134,18 +188,150 @@ func main() {
 		}
 	}
 
+	runOptions := controllerRunOptions{
+		kubeClient:           kubeClient,
+		extClient:            extClient,
+		machineClient:        machineClient,
+		sshKeyPair:           key,
+		metrics:              NewMachineControllerMetrics(),
+		clusterDNSIPs:        ips,
+		leaderElectionClient: leaderElectionClient,
+		nodeInformer:         nodeInformer,
+		configMapInformer:    configMapInformer,
+		machineInformer:      machineInformer,
+		kubeconfigProvider:   kubeconfigProvider,
+	}
+
+	startUtilHttpServerOrDie(kubeClient, kubeconfigProvider, stopCh)
+	startControllerViaLeaderElectionOrDie(runOptions)
+}
+
+// startControllerViaLeaderElectionOrDie starts machine controller only if a proper lock was acquired.
+// This essentially means that we can have multiple instances and at the same time only one is operational.
+// The program terminates when the leadership was lost.
+func startControllerViaLeaderElectionOrDie(runOptions controllerRunOptions) {
+	id, err := os.Hostname()
+	if err != nil {
+		glog.Fatalf("error getting hostname: %s", err.Error())
+	}
+	// add a seed to the id, so that two processes on the same host don't accidentally both become active
+	id = id + "_" + string(uuid.NewUUID())
+
+	rl := resourcelock.EndpointsLock{
+		EndpointsMeta: metav1.ObjectMeta{
+			Namespace: defaultLeaderElectionNamespace,
+			Name:      controllerName,
+		},
+		Client: runOptions.leaderElectionClient.CoreV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity:      id + fmt.Sprintf("-%s", controllerName),
+			EventRecorder: createRecorder(runOptions.kubeClient),
+		},
+	}
+
+	run := func(stopCh <-chan struct{}) {
+		machineController := controller.NewMachineController(
+			runOptions.kubeClient,
+			runOptions.machineClient,
+			runOptions.nodeInformer,
+			runOptions.configMapInformer,
+			runOptions.machineInformer,
+			runOptions.sshKeyPair,
+			runOptions.clusterDNSIPs,
+			controller.MetricsCollection{
+				Machines:            runOptions.metrics.Machines,
+				Workers:             runOptions.metrics.Workers,
+				Errors:              runOptions.metrics.Errors,
+				Nodes:               runOptions.metrics.Nodes,
+				ControllerOperation: runOptions.metrics.ControllerOperation,
+				NodeJoinDuration:    runOptions.metrics.NodeJoinDuration,
+			},
+			runOptions.kubeconfigProvider)
+
+		if err = machineController.Run(workerCount, stopCh); err != nil {
+			glog.Fatalf("error running controller: %v", err)
+		}
+	}
+
+	leaderelection.RunOrDie(leaderelection.LeaderElectionConfig{
+		Lock:          &rl,
+		LeaseDuration: defaultLeaderElectionLeaseDuration,
+		RenewDeadline: defaultLeaderElectionRenewDeadline,
+		RetryPeriod:   defaultLeaderElectionRetryPeriod,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: run,
+			OnStoppedLeading: func() {
+				glog.Infof("leaderelection lost, closing the app")
```

Comment:
should this be a `fatal`? or is there any other callback which exists?

Reply:
yes, should be `fatal` - thanks! To the best of my knowledge only `OnStoppedLeading` exits.